crates/code_generation/src/cyclers.rs

```diff
@@ -638,22 +835,63 @@ fn generate_perception_cycler_updates(cyclers: &Cyclers) -> TokenStream {
 fn generate_node_execution(
     node: &Node,
     cycler: &Cycler,
-    recording_generation: RecordingGeneration,
+    node_type: NodeType,
+    mode: Execution,
+) -> TokenStream {
+    match (node_type, mode) {
+        (NodeType::Setup, Execution::Run) => {
+            let write_main_outputs_from_node =
+                generate_write_main_outputs_from_node(node, cycler, mode);
+            let record_main_outputs = generate_record_main_outputs(node);
+            quote! {
+                #write_main_outputs_from_node
+                #record_main_outputs
+            }
+        }
+        (NodeType::Cycle, Execution::Run) => {
+            let record_node_state = generate_record_node_state(node);
+            let write_main_outputs_from_node =
+                generate_write_main_outputs_from_node(node, cycler, mode);
+            quote! {
+                #record_node_state
+                #write_main_outputs_from_node
+            }
+        }
+        (NodeType::Setup, Execution::Replay) => {
+            let write_main_outputs_from_frame = generate_write_main_outputs_from_frame(node);
+            quote! {
+                #write_main_outputs_from_frame
+            }
+        }
+        (NodeType::Cycle, Execution::Replay) => {
+            let restore_node_state = generate_restore_node_state(node);
+            let write_main_outputs_from_node =
+                generate_write_main_outputs_from_node(node, cycler, mode);
+            quote! {
+                #restore_node_state
+                #write_main_outputs_from_node
+            }
+        }
+        (_, Execution::None) => panic!("unexpected Execution::Mode"),
+    }
+}
+
+fn generate_write_main_outputs_from_node(
```

Comment:
I will think about this... :hourglass_flowing_sand:

Reply:
The previous call tree is

- `generate_node_execution`
    - `generate_record_node_state`
    - `generate_write_main_outputs_from_node`
        - `generate_database_updates`
        - `generate_database_updates_from_defaults`
    - `generate_write_main_outputs_from_frame`
    - `generate_record_main_outputs`

I would argue that the upper level `generate_node_execution` is good enough from the viewport of the caller because it is short and hides recording implementation details.

The `generate_write_main_outputs_from_node` can be improved by adding that it also executes the node: `generate_execute_node_and_write_main_outputs`

Similar, `generate_write_main_outputs_from_frame` has been renamed to `generate_deserialize_frame_and_write_main_outputs`.

I would argue that the most inner functions are fine because they are more narrow then compared to `generate_execute_node_and_write_main_outputs`. But I adapted the `write_main_outputs` now.

This leads to the new naming:

- `generate_node_execution`
    - `generate_record_node_state`
    - `generate_execute_node_and_write_main_outputs`
        - `generate_write_main_outputs`
        - `generate_write_main_outputs_from_defaults`
    - `generate_deserialize_frame_and_write_main_outputs`
    - `generate_record_main_outputs`